File: src/ur_script_sender_cpp/src/send_traj_action.cpp
#include <chrono>
#include <string>
#include <vector>
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>

#include <control_msgs/action/follow_joint_trajectory.hpp>
#include <trajectory_msgs/msg/joint_trajectory_point.hpp>

#include <yaml-cpp/yaml.h>
#include <ament_index_cpp/get_package_share_directory.hpp>
#include <filesystem>

using FollowJT = control_msgs::action::FollowJointTrajectory;
using GoalHandleFollowJT = rclcpp_action::ClientGoalHandle<FollowJT>;

static std::string get_action_name(int argc, char** argv) {
  // argv[2] optional
  if (argc >= 3) return argv[2];
  return "/scaled_joint_trajectory_controller/follow_joint_trajectory";
}

class TrajActionSender : public rclcpp::Node {
public:
  explicit TrajActionSender(const std::string& action_name)
  : Node("ur_traj_action_sender_cpp"), action_name_(action_name) {
    client_ = rclcpp_action::create_client<FollowJT>(this, action_name_);
  }

  void send_from_yaml(const std::string& yaml_path) {
    YAML::Node root = YAML::LoadFile(yaml_path);

    auto joint_names = root["joint_names"].as<std::vector<std::string>>();
    YAML::Node points = root["points"];
    if (!points || !points.IsSequence() || points.size() == 0) {
      throw std::runtime_error("YAML 'points' must be a non-empty sequence.");
    }

    FollowJT::Goal goal;
    goal.trajectory.joint_names = joint_names;

    double last_t = -1.0;
    for (const auto& p : points) {
      double t = p["t"].as<double>();
      if (t < 0.0 || t < last_t) {
        throw std::runtime_error("Each point.t must be non-negative and non-decreasing.");
      }
      last_t = t;

      trajectory_msgs::msg::JointTrajectoryPoint jp;
      jp.positions = p["positions"].as<std::vector<double>>();

      if (p["velocities"]) jp.velocities = p["velocities"].as<std::vector<double>>();
      if (p["accelerations"]) jp.accelerations = p["accelerations"].as<std::vector<double>>();

      // time_from_start
      int32_t sec = static_cast<int32_t>(t);
      uint32_t nanosec = static_cast<uint32_t>((t - sec) * 1e9);
      jp.time_from_start.sec = sec;
      jp.time_from_start.nanosec = nanosec;

      goal.trajectory.points.push_back(jp);
    }

    RCLCPP_INFO(get_logger(), "Waiting for action server: %s", action_name_.c_str());
    if (!client_->wait_for_action_server(std::chrono::seconds(5))) {
      throw std::runtime_error("Action server not available. Check controller/action name.");
    }

    rclcpp_action::Client<FollowJT>::SendGoalOptions opts;
    opts.feedback_callback =
      [this](GoalHandleFollowJT::SharedPtr, const std::shared_ptr<const FollowJT::Feedback> /*fb*/) {
        // 필요하면 feedback 출력
      };

    opts.result_callback =
      [this](const GoalHandleFollowJT::WrappedResult & result) {
        const auto& res = result.result;
        RCLCPP_INFO(this->get_logger(), "Done. error_code=%d, error_string='%s'",
          res->error_code, res->error_string.c_str());
        rclcpp::shutdown();
      };

    RCLCPP_INFO(get_logger(), "Sending trajectory with %zu points...", goal.trajectory.points.size());
    auto future = client_->async_send_goal(goal, opts);

    // goal accept/reject는 future 완료 후 확인 가능
    // 여기서는 spin으로 진행
  }

private:
  std::string action_name_;
  rclcpp_action::Client<FollowJT>::SharedPtr client_;
};

int main(int argc, char** argv) {
  if (argc < 2) {
    std::cerr << "Usage: send_traj_action <path.yaml> [action_name]\n";
    return 2;
  }

  std::string yaml_path = argv[1];
  std::string action_name = get_action_name(argc, argv);

  rclcpp::init(argc, argv);
  auto node = std::make_shared<TrajActionSender>(action_name);

  try {
    node->send_from_yaml(yaml_path);
  } catch (const std::exception& e) {
    RCLCPP_ERROR(node->get_logger(), "Error: %s", e.what());
    rclcpp::shutdown();
    return 1;
  }

  rclcpp::spin(node);
  return 0;
}

File: src/ur_script_sender_py/ur_script_sender_py/send_traj_action.py
import sys
import time
import yaml
import os

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient

from control_msgs.action import FollowJointTrajectory
from trajectory_msgs.msg import JointTrajectoryPoint
from ament_index_python.packages import get_package_share_directory

DEFAULT_ACTION = "/scaled_joint_trajectory_controller/follow_joint_trajectory"


def default_yaml_path():
    share = get_package_share_directory("ur_script_sender_py")
    return os.path.join(share, "paths", "demo_joint_path.yaml")


class TrajActionSender(Node):
    def __init__(self, action_name: str):
        super().__init__("ur_traj_action_sender")
        self._client = ActionClient(self, FollowJointTrajectory, action_name)

    def send_from_yaml(self, yaml_path: str):
        with open(yaml_path, "r") as f:
            data = yaml.safe_load(f)

        joint_names = data["joint_names"]
        pts = data["points"]

        n = len(joint_names)
        for i, p in enumerate(pts):
            if len(p["positions"]) != n:
                raise ValueError(
                    f"points[{i}].positions length {len(p['positions'])} != joint_names length {n}"
                )

        goal = FollowJointTrajectory.Goal()
        goal.trajectory.joint_names = joint_names

        last_t = -1.0
        for p in pts:
            t = float(p["t"])
            if t < 0 or t < last_t:
                raise ValueError("Each point.t must be non-negative and non-decreasing.")
            last_t = t

            jp = JointTrajectoryPoint()
            jp.positions = [float(x) for x in p["positions"]]

            if "velocities" in p:
                jp.velocities = [float(x) for x in p["velocities"]]
            if "accelerations" in p:
                jp.accelerations = [float(x) for x in p["accelerations"]]

            sec = int(t)
            nanosec = int((t - sec) * 1e9)
            jp.time_from_start.sec = sec
            jp.time_from_start.nanosec = nanosec

            goal.trajectory.points.append(jp)

        self.get_logger().info(f"Waiting for action server: {DEFAULT_ACTION} ...")
        if not self._client.wait_for_server(timeout_sec=5.0):
            raise RuntimeError("Action server not available. Check controller/action name.")

        time.sleep(0.1)

        self.get_logger().info(f"Sending {len(goal.trajectory.points)} points... yaml={yaml_path}")
        send_future = self._client.send_goal_async(goal, feedback_callback=self._feedback_cb)
        rclpy.spin_until_future_complete(self, send_future)

        goal_handle = send_future.result()
        if not goal_handle.accepted:
            raise RuntimeError("Goal rejected by controller.")

        self.get_logger().info("Goal accepted. Waiting for result...")
        result_future = goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, result_future)

        res = result_future.result().result
        self.get_logger().info(f"Done. error_code={res.error_code}, error_string='{res.error_string}'")

    def _feedback_cb(self, fb_msg):
        fb = fb_msg.feedback
        t = fb.actual.time_from_start
        self.get_logger().info(f"feedback actual t={t.sec}.{t.nanosec:09d}")


def main():
    # 사용법:
    # 1) ros2 run ... send_traj_action               -> share/paths/demo_joint_path.yaml 사용
    # 2) ros2 run ... send_traj_action <yaml>        -> 지정 yaml 사용
    # 3) ros2 run ... send_traj_action <yaml> <action_name>
    if len(sys.argv) >= 2:
        yaml_path = sys.argv[1]
    else:
        yaml_path = default_yaml_path()

    action_name = sys.argv[2] if len(sys.argv) >= 3 else DEFAULT_ACTION

    # 만약 yaml_path가 "paths/xxx.yaml"처럼 상대경로면 현재 cwd 기준이라 깨질 수 있음.
    # 필요하면 여기서 절대경로로 강제할 수도 있음:
    # yaml_path = os.path.abspath(yaml_path)

    rclpy.init()
    node = TrajActionSender(action_name)
    try:
        node.send_from_yaml(yaml_path)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
